### PR TITLE
injected options

### PR DESCRIPTION
**Pull request by juliuskoval**
Addresses https://github.com/juliuskoval/NLog.Targets.OpenTelemetryProtocol/issues/13.

@snakefoot would you mind taking a look at this please?

@TheXenocide is this what you had in mind?

### CODE REVIEW

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs [hunk 1]
```diff
@@ -34,6 +36,10 @@ public class OtlpTarget : TargetWithContext
 
         public bool IncludeFormattedMessage { get; set; }
 
+        public bool UseInjectedOptions { get; set; }
```
Comment (snakefoot) [inline NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs:39]: Would remove this option, and only rely on whether the resolved option-name is not empty string.

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs [hunk 2]
```diff
@@ -34,6 +36,10 @@ public class OtlpTarget : TargetWithContext
...
+
+        public Layout InjectedOptionsName { get; set; }
```
Comment (snakefoot) [inline NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs:41]: Maybe rename this to `ResolveOptionsFromName`

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs [hunk 3]
```diff
@@ -68,14 +74,14 @@ protected override void InitializeTarget()
...
-            var endpoint = RenderLogEvent(Endpoint, LogEventInfo.CreateNullEvent());
-            var useHttp = RenderLogEvent(UseHttp, LogEventInfo.CreateNullEvent());
-            var headers = RenderLogEvent(Headers, LogEventInfo.CreateNullEvent());
+
             var maxQueueSize = RenderLogEvent(MaxQueueSize, LogEventInfo.CreateNullEvent(), 2048);
             var maxExportBatchSize = RenderLogEvent(MaxExportBatchSize, LogEventInfo.CreateNullEvent(), 512);
             var scheduledDelayMilliseconds = RenderLogEvent(ScheduledDelayMilliseconds, LogEventInfo.CreateNullEvent(), 5000);
 
-            _processor = CreateProcessor(endpoint, useHttp, headers, maxQueueSize, maxExportBatchSize, scheduledDelayMilliseconds);
+            var options = UseInjectedOptions ? GetInjectedOtlpExporterOptions() : GetOtlpExporterOptions();
```
Comment (snakefoot) [inline NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs:82]: Maybe make it like this:
```c#
var options = ResolveOtlpExporterOptionsFromName() ?? GetOtlpExporterOptions();
```
Where `ResolveOtlpExporterOptionsFromName` returns `null`, when `ResolveOptionsFromName` renders into null or empty-string.

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs (hunk 1 above)
Comment (juliuskoval) [reply NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs]: done

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs [hunk 4]
```diff
@@ -91,42 +95,46 @@ protected override void InitializeTarget()
...
-            if (!result)
-                InternalLogger.Info("OtlpTarget(Name={0}) - Shutdown OpenTelemetry BatchProcessor unsuccesful");
+            if (string.IsNullOrEmpty(resolvedOptionsName))
+            {
+                return null;
+            }
 
-            base.CloseTarget();
-        }
+            using IServiceScope scope = this.ResolveService<IServiceScopeFactory>().CreateScope();
```
Comment (snakefoot) [inline NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs:107]: Maybe add some logging like this:
```c#
NLog.Common.InternalLogger.Debug("{0} - Resolves OtlpExporterOptions from name '{1}'", this,  resolvedOptionsNam);
```

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs [hunk 5]
```diff
@@ -173,6 +181,27 @@ private ResourceBuilder CreateResourceBuilder()
...
+            var logProvider = _loggerProvider;
+            _loggerProvider = null;
+            logProvider?.Dispose();
+            _loggers.Clear();
+
+            var processor = _processor;
+            _processor = null;
+            var result = processor?.Shutdown(1000) ?? true;
+            if (!result)
+                InternalLogger.Info("OtlpTarget(Name={0}) - Shutdown OpenTelemetry BatchProcessor unsuccesful");
```
Comment (snakefoot) [inline NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs:195]: You are missing `{0}` as parameter. Maybe something like this:
```c#
InternalLogger.Info("{0} - Shutdown OpenTelemetry BatchProcessor unsuccesful", this);
```

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs (hunk 4 above)
Comment (juliuskoval) [reply NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs:107]: done

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs (hunk 5 above)
Comment (juliuskoval) [reply NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs]: fixed

File: NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs [hunk 6]
```diff
@@ -173,6 +183,27 @@ private ResourceBuilder CreateResourceBuilder()
...
+            var logProvider = _loggerProvider;
+            _loggerProvider = null;
+            logProvider?.Dispose();
+            _loggers.Clear();
+
+            var processor = _processor;
+            _processor = null;
+            var result = processor?.Shutdown(1000) ?? true;
+            if (!result)
+                InternalLogger.Info("OtlpTarget(Name={0}) - Shutdown OpenTelemetry BatchProcessor unsuccessful", this);
```
Comment (snakefoot) [inline NLog.Targets.OpenTelemetryProtocol/OtlpTarget.cs:197]: Either choose between:
```
InternalLogger.Info("{0} - Shutdown OpenTelemetry BatchProcessor unsuccessful", this);
```
Or this:
```
InternalLogger.Info("OtlpTarget(Name={0}) - Shutdown OpenTelemetry BatchProcessor unsuccessful", Name);
```

**Comment by TheXenocide** [comment]
Yes indeed, looks good to me. I'm not sure about the applicability of IOptionsSnapshot (vs IOptionsMonitor) since I haven't read through enough of the code to know if it will pick up updates in the config at runtime, but that functionality isn't so important to me compared to being able to leverage the existing configuration pattern which this appears to satisfy from what I can see.

**Pull request merged**